### PR TITLE
xbo

### PR DESCRIPTION
**Pull request by pperidont**

### CODE REVIEW

File: libs/directx/haxelib.json [hunk 1]
```diff
@@ -6,5 +6,4 @@
 	"description" : "DirectX support for Haxe/HL.",
 	"version" : "1.4.0",
 	"releasenote" : "",
-	"dependencies": { "hlopenal" : "" }
```
Comment (ncannasse) [inline libs/directx/haxelib.json]: Please no :)

File: libs/directx/directx.cpp [hunk 2]
```diff
@@ -132,11 +144,13 @@ HL_PRIM void HL_NAME(clear_color)( dx_pointer *rt, double r, double g, double b,
 	color[1] = (float)g;
 	color[2] = (float)b;
 	color[3] = (float)a;
-	driver->context->ClearRenderTargetView((ID3D11RenderTargetView*)rt,color);
+	ID3D11RenderTargetView* renderTarget = (ID3D11RenderTargetView*)rt;
+	driver->context->ClearRenderTargetView(renderTarget,color);
+	driver->context->OMSetRenderTargets(1, &renderTarget, depthBuffer);
```
Comment (ncannasse) [inline libs/directx/directx.cpp]: This is a violation of the API rules : this should be handled at the driver level !

File: libs/directx/directx.cpp [hunk 3]
```diff
@@ -67,18 +65,25 @@ HL_PRIM dx_driver *HL_NAME(create)( HWND window, int format, int flags, int rest
 	desc.BufferDesc.Format = (DXGI_FORMAT)format;
 	desc.SampleDesc.Count = 1; // NO AA for now
 	desc.BufferUsage = DXGI_USAGE_RENDER_TARGET_OUTPUT;
-	desc.BufferCount = 1;
+	desc.BufferCount = 2;
```
Comment (ncannasse) [inline libs/directx/directx.cpp]: No triple buffering please

File: libs/directx/directx.cpp [hunk 4]
```diff
@@ -26,9 +23,10 @@ template <typename T> class dx_struct {
 typedef ID3D11Resource dx_resource;
 typedef ID3D11DeviceChild dx_pointer;
 
-static dx_driver *driver = NULL;
+dx_driver *driver = NULL;
```
Comment (ncannasse) [inline libs/directx/directx.cpp]: a static + a get function is better for such a normally named variable

File: libs/directx/directx.cpp [hunk 5]
```diff
@@ -132,11 +144,13 @@ HL_PRIM void HL_NAME(clear_color)( dx_pointer *rt, double r, double g, double b,
...
 	color[3] = (float)a;
-	driver->context->ClearRenderTargetView((ID3D11RenderTargetView*)rt,color);
+	ID3D11RenderTargetView* renderTarget = (ID3D11RenderTargetView*)rt;
+	driver->context->ClearRenderTargetView(renderTarget,color);
+	driver->context->OMSetRenderTargets(1, &renderTarget, depthBuffer);
 }
 
 HL_PRIM void HL_NAME(present)( int interval, int flags ) {
-	driver->swapchain->Present(interval,flags);
+	HRESULT res = driver->swapchain->Present(interval, flags); assert(res == S_OK);
```
Comment (ncannasse) [inline libs/directx/directx.cpp]: Use DXERR instead?

File: libs/directx/directx.cpp [hunk 6]
```diff
@@ -26,9 +23,10 @@ template <typename T> class dx_struct {
 typedef ID3D11Resource dx_resource;
 typedef ID3D11DeviceChild dx_pointer;
 
-static dx_driver *driver = NULL;
+dx_driver *driver = NULL;
 static IDXGIFactory *factory = NULL;
 
+C_FUNCTION_BEGIN
```
Comment (ncannasse) [inline libs/directx/directx.cpp]: this should not be necessary given that HL_IMPL defines extern "C" ?

File: src/hl.h [hunk 7]
```diff
@@ -29,8 +29,12 @@
 
 #define HL_VERSION	0x140
 
-#ifdef _WIN32
+//WIN32 is defined in both windows and xbo
+#if defined(_WIN32)
 #	define HL_WIN
+#   ifndef _DURANGO
+#       define HL_WIN_DESKTOP
+#   endif
```
Comment (ncannasse) [inline src/hl.h]: Tabs please + remove comment

File: src/hl.h [hunk 8]
```diff
@@ -65,7 +69,11 @@
 #	define HL_NX
 #endif
 
-#if defined(HL_PS) || defined(HL_NX)
+#ifdef _DURANGO
+# define HL_XBO
```
Comment (ncannasse) [inline src/hl.h]: tabs

File: src/hl.h [hunk 9]
```diff
@@ -183,7 +191,11 @@ typedef unsigned long long uint64;
 // -------------- UNICODE -----------------------------------
 
 #if defined(HL_WIN) && !defined(HL_LLVM)
-#	include <windows.h>
+#ifdef HL_WIN_DESKTOP
+#   include <Windows.h>
```
Comment (ncannasse) [inline src/hl.h]: tabs

File: src/hlc.h [hunk 10]
```diff
@@ -44,17 +44,18 @@
...
+#	pragma warning(disable:4102) // unreferenced label
+#	pragma warning(disable:4700) // uninitialized local variable used
 #	pragma warning(disable:4703) // potentially uninitialized local
-#	pragma warning(disable:4723) // potential divide by 0
-#	pragma warning(disable:4715) // control paths must return a value
-#	pragma warning(disable:4716) // must return a value (ends with throw)
+#	pragma warning(disable:4723) // potential divide by 0
+#	pragma warning(disable:4715) // control paths must return a value
+#	pragma warning(disable:4716) // must return a value (ends with throw)
+#	pragma warning(disable:4701) // potentially uninitialized local variable
```
Comment (ncannasse) [inline src/hlc.h]: what's going on here exactly?

File: src/std/debug.c [hunk 11]
```diff
@@ -21,7 +21,8 @@
  */
 #include <hl.h>
 
-#ifdef HL_WIN
+
+#ifdef HL_WIN_DESKTOP
```
Comment (ncannasse) [inline src/std/debug.c]: better revert everything from this file and not include it in xbo project no?

File: src/std/file.c [hunk 12]
```diff
@@ -25,7 +25,11 @@
 #	include <posix/posix.h>
 #endif
 #ifdef HL_WIN
+#ifdef HL_WIN_DESKTOP
 #	include <windows.h>
+#else
+#   include<xdk.h>
```
Comment (ncannasse) [inline src/std/file.c]: tabs

File: src/hlc.h (hunk 10 above)
Comment (pperidont) [reply src/hlc.h]: This file had mixed line endings. 
We only added `#pragma warning(disable:4701) // potentially uninitialized local variable`

File: src/std/math.c [hunk 13]
```diff
@@ -36,7 +36,11 @@ HL_PRIM double hl_math_abs( double a ) {
 }
 
 HL_PRIM bool hl_math_isnan( double a ) {
-	return a != a;
+#ifdef HL_WIN
+	return isnan(a);
+#else
+	return a != a; //does not work on some platforms
+#endif
```
Comment (ncannasse) [inline src/std/math.c]: Is it xbo related?

**Pull request merged**